app/src/main/java/com/example/bottomnav/Parse.java

```diff
@@ -51,75 +53,167 @@ public Parse(char[] code) throws Exception {
+            case 5:
+                return "double";
+            default:
+                return null;
+        }
     }
 
-    public Optional<DataDecoder> getDecoder(Integer canID) {
-        return getDecoder(canIdToName.get(canID));
+    /**
+     * Creates and stores a Primitive decoder if declaration is a Const
+     * @param declaration
+     * @throws Exception
+     */
+    private void storeConstDecoder(CPPASTSimpleDeclaration declaration)
+            throws ExpansionOverlapsBoundaryException {
+        Optional<VariableContents> data = parseStatement(declaration).map(this::isAppropriateConst);
+        if (data.isPresent()) {
+            VariableContents variableContents = data.get();
+            Optional<DataDecoder> decoder = PrimitiveDecoder.create(variableContents);
+            if (decoder.isPresent()){
+                decoderRepo.put(variableContents.name, decoder.get());
+                idToName.put(Integer.decode(variableContents.value), variableContents.name);
+            }
+        }
     }
 
-    public VariableContents getConstContents(String name) {
-        return ((PrimitiveDecoder) decoderRepo.get(name).get()).contents;
+    /**
+     * Iterates over declarations of struct, makes array variableContents, then creates and stores
+     * struct decoder with array of variableContents
+     */
+    private void storeStructDecoder(IASTDeclaration[] declarations, String name)
+            throws ExpansionOverlapsBoundaryException {
+        ArrayList<VariableContents> variables = new ArrayList<>();
+        for (IASTDeclaration element : declarations) {
+            CPPASTSimpleDeclaration declaration = (CPPASTSimpleDeclaration) element;
+            Optional<VariableContents> data = parseStatement(declaration).map(this::isAppropriateMember);
+            if (data.isPresent()) {
+                variables.add(data.get());
+            }
+        }
+        Optional<DataDecoder> structDecoder = StructDecoder.create(variables);
+        if (structDecoder.isPresent()){
+            decoderRepo.put(name, structDecoder.get());
+        }
+    }
+
+    public Optional<DataDecoder> getDecoder(String name) {
+        if (!decoderRepo.containsKey(name)) {
+            return Optional.empty();
+        }
+        return Optional.of(decoderRepo.get(name));
     }
 
-    public ArrayList<VariableContents> getStructContents(String name) {
-        return ((StructDecoder) getDecoder(name).get()).variables;
+    public Optional<DataDecoder> getDecoder(Integer canId) {
+        return getDecoder(idToName.get(canId));
+    }
+
+    /**
+     * The decoders return a string format of 'NAME: Value, NAME: Value, ...'
+     * Therefore, parseStringResults returns an array of Decoded objects that allows user to iterate
+     * over and use functions like getName and getValue in for each Decoded object
+     *
+     * @return
+     */
+    public ArrayList<DecodedContents> parseDecodedString(String resultOfDecoder) {
```

Comment:
Packing structured data into a string then reverse-engineering that string into structured data is kind of a red flag. Can this be done more directly?

Reply:
I believe that the decoders themselves should only return an optional of a string.  I thought about returning an optional of an object where variables would be name and value.  But it gets a little complicated because struct decoder uses primitive decoder since I want them to work universally: Decode(payload) => optional(string)

Reply:
parseDecoded looks like it's only used in test cases? If that's the case, consider moving it to the test code as a test helper function.

In general, I'd say that structured data -> string -> decode back into structured data is a major red flag and can be a longer term software maintenance headache.